opal/mca/common/ucx/common_ucx_wpool.c

```diff
@@ -740,38 +789,42 @@ opal_common_ucx_tlocal_fetch_spath(opal_common_ucx_wpmem_t *mem, int target)
     _mem_record_t *mem_rec = NULL;
     opal_common_ucx_winfo_t *winfo = NULL;
     ucp_ep_h ep;
-    int rc = OPAL_SUCCESS;
+    int rc = OPAL_SUCCESS, proc_vpid;
 
     ctx_rec = _tlocal_get_ctx_rec(mem->ctx->tls_key);
     if (OPAL_UNLIKELY(!ctx_rec)) {
-        ctx_rec = _tlocal_add_ctx_rec(mem->ctx);
+        ctx_rec = _tlocal_add_ctx_rec(mem->ctx, mem->comm_size);
         if (NULL == ctx_rec) {
             return OPAL_ERR_OUT_OF_RESOURCE;
         }
     }
     winfo = ctx_rec->winfo;
+    
+    proc_vpid = ctx_rec->gctx->get_proc_vpid_func(mem->metadata, target);
 
     /* Obtain the endpoint */
-    if (OPAL_UNLIKELY(NULL == winfo->endpoints[target])) {
-        rc = _tlocal_ctx_connect(ctx_rec, target);
+    if (OPAL_UNLIKELY(NULL == winfo->endpoints[proc_vpid])) {
+        rc = _tlocal_ctx_connect(mem, ctx_rec, target);
         if (rc != OPAL_SUCCESS) {
             return rc;
         }
     }
-    ep = winfo->endpoints[target];
+    ep = winfo->endpoints[proc_vpid];
 
     /* Obtain the memory region info */
     mem_rec = _tlocal_add_mem_rec(mem, ctx_rec);
```

Comment:
Fetch it first to see if it was already allocated.
May want to pass mem_rec here